src/lib/optimizer/strategy/projection_combination_rule.cpp

```diff
@@ -0,0 +1,106 @@
+#include "projection_combination_rule.hpp"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+
+namespace opossum {
+
+std::string ProjectionCombinationRule::name() const { return "Projection Combination Rule"; }
+
+bool ProjectionCombinationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  if (node->type() != LQPNodeType::Projection) {
+    return _apply_to_children(node);
+  }
+
+  std::vector<std::shared_ptr<ProjectionNode>> projection_nodes;
+
+  // Gather subsequent ProjectionNodes
+  auto current_node = node;
+  while (current_node->type() == LQPNodeType::Projection) {
+    projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(current_node));
+
+    // If current node has multiple or no children, end the combination here
+    if (current_node->right_child() || !current_node->left_child()) {
+      break;
+    }
+
+    current_node = current_node->left_child();
+
+    if (current_node->parents().size() > 1) {
+      break;
+    }
+  }
+
+  if (projection_nodes.size() > 1) {
+    const auto combined_projection_node = _combine_projections(projection_nodes);
+    _apply_to_children(combined_projection_node);
+    return true;
+  }
+
+  return _apply_to_children(node);
+}
+
+std::shared_ptr<ProjectionNode> ProjectionCombinationRule::_combine_projections(
+    std::vector<std::shared_ptr<ProjectionNode>>& projections) const {
+  // Store original child and parents
+  auto left_child = projections.back()->left_child();
+  auto right_child = projections.back()->right_child();
+  const auto parents = projections.front()->parents();
+  const auto child_sides = projections.front()->get_child_sides();
+
+  auto column_expressions = std::vector<std::shared_ptr<LQPExpression>>();
+  for (const auto expression : projections.front()->column_expressions()) {
+    // If the expression is no reference to another column, we can just add it.
+    if (expression->type() != ExpressionType::Column) {
+      column_expressions.push_back(expression);
+      continue;
+    }
+
+    // If the expression IS a reference to another column,
+    // we have to check if the referenced column is part of one of the other ProjectionNodes that we are combining,
+    // and then instead add the referenced column here.
+    auto column_reference_replaced = false;
```

Comment:
You could (imo: should ;) ) really take advantage of what a LQPColumnReference IS here: a LQPNode and a ColumnID. Don't blindly search for the LQPColumnReference, search for the LQPNode among the projection_nodes (possibly using a map) and then just use the ColumnID to offset into its projection_expressions. This would be much more elegant and faster. (Remember some TPCH query happily go up to ~50 rows!)

Reply:
Done, thanks!